File: gsh/gsh_kreatao/report/unmarked_attendance_days/unmarked_attendance_days.py
import json
import frappe
from frappe.utils import getdate, add_days

OFF_SHIFT_TYPES = ["Weekly Off", "Night Off"]

def execute(filters=None):
    filters = filters or {}
    from_date = getdate(filters.get("from_date"))
    to_date = getdate(filters.get("to_date"))
    department = filters.get("department")

    columns = [
        {"label": "Employee", "fieldname": "employee", "fieldtype": "Link", "options": "Employee", "width": 150},
        {"label": "Employee Name", "fieldname": "employee_name", "fieldtype": "Data", "width": 180},
        {"label": "Department", "fieldname": "department", "fieldtype": "Link", "options": "Department", "width": 150},
        {"label": "Total Unmarked Days", "fieldname": "total_unmarked", "fieldtype": "HTML", "width": 160},
        {"label": "unmarked_dates_json", "fieldname": "unmarked_dates_json", "fieldtype": "Data", "width": 0, "hidden": 1},
    ]

    emp_filters = {"status": "Active"}
    if department:
        emp_filters["department"] = department

    employees = frappe.get_all(
        "Employee",
        filters=emp_filters,
        fields=["name", "employee_name", "department", "date_of_joining", "relieving_date", "holiday_list"]
    )
    employee_ids = [e.name for e in employees]

    marked_by_date = {}
    d = from_date
    while d <= to_date:
        marked_by_date[d] = set(frappe.get_all("Attendance", filters={"attendance_date": d}, pluck="employee"))
        d = add_days(d, 1)

    # WO/NO from Shift Assignment
    shift_assignments = frappe.get_all(
        "Shift Assignment",
        filters={
            "employee": ["in", employee_ids],
            "shift_type": ["in", OFF_SHIFT_TYPES],
            "docstatus": 1,
            "start_date": ["<=", to_date],
        },
        fields=["employee", "shift_type", "start_date", "end_date"]
    )

    off_dates_by_employee = {}
    for sa in shift_assignments:
        start = sa.start_date
        end = sa.end_date or sa.start_date
        if end < from_date:
            continue
        d2 = max(start, from_date)
        end_clamped = min(end, to_date)
        emp_off = off_dates_by_employee.setdefault(sa.employee, set())
        while d2 <= end_clamped:
            emp_off.add(d2)
            d2 = add_days(d2, 1)

    # Holiday List dates per distinct holiday list used by these employees
    holiday_lists = list({e.holiday_list for e in employees if e.holiday_list})
    holidays_by_list = {}
    if holiday_lists:
        all_holidays = frappe.get_all(
            "Holiday",
            filters={
                "parent": ["in", holiday_lists],
                "holiday_date": ["between", [from_date, to_date]],
            },
            fields=["parent", "holiday_date"]
        )
        for h in all_holidays:
            holidays_by_list.setdefault(h.parent, set()).add(h.holiday_date)

    data = []
    for emp in employees:
        emp_off_dates = off_dates_by_employee.get(emp.name, set())
        emp_holiday_dates = holidays_by_list.get(emp.holiday_list, set()) if emp.holiday_list else set()

        unmarked_iso = []
        d = from_date
        while d <= to_date:
            if emp.date_of_joining and d < emp.date_of_joining:
                d = add_days(d, 1)
                continue
            if emp.relieving_date and d > emp.relieving_date:
                d = add_days(d, 1)
                continue
            if d in emp_off_dates or d in emp_holiday_dates:
                d = add_days(d, 1)
                continue
            if emp.name not in marked_by_date[d]:
                unmarked_iso.append(str(d))
            d = add_days(d, 1)

        if unmarked_iso:
            count = len(unmarked_iso)
            total_html = (
                f"<a href='#' class='unmarked-count-link' "
                f"data-employee='{emp.name}' "
                f"data-employee-name='{frappe.utils.escape_html(emp.employee_name)}' "
                f"style='font-weight:600; color:#2490ef; text-decoration:underline; cursor:pointer;'>"
                f"{count}</a>"
            )
            data.append({
                "employee": emp.name,
                "employee_name": emp.employee_name,
                "department": emp.department,
                "total_unmarked": total_html,
                "unmarked_dates_json": json.dumps(unmarked_iso),
            })

    data.sort(key=lambda x: len(json.loads(x["unmarked_dates_json"])), reverse=True)

    return columns, data

File: gsh/gsh_kreatao/report/unmarked_attendance_days/unmarked_attendance_days.js
frappe.query_reports["Unmarked Attendance Days"] = {
"filters": [
{
"fieldname": "from_date",
"label": "From Date",
"fieldtype": "Date",
"default": frappe.datetime.month_start(),
"reqd": 1
},
{
"fieldname": "to_date",
"label": "To Date",
"fieldtype": "Date",
"default": frappe.datetime.month_end(),
"reqd": 1
},
{
"fieldname": "department",
"label": "Department",
"fieldtype": "Link",
"options": "Department"
}
],
"datatable": {
"dynamicRowHeight": false
}
};

// delegated click handler bound once on document - works regardless of render hook availability
$(document).off("click", ".unmarked-count-link").on("click", ".unmarked-count-link", function(e) {
e.preventDefault();
const employee = $(this).data("employee");
const employee_name = $(this).data("employee-name");

const report_data = (frappe.query_report && frappe.query_report.data) || [];
const row = report_data.find(r => r.employee === employee);
if (!row) {
frappe.msgprint("Could not find data for this employee.");
return;
}

let dates = [];
try {
dates = JSON.parse(row.unmarked_dates_json || "[]");
} catch (err) {
dates = [];
}
const date_set = new Set(dates);

const months = {};
dates.forEach(d => {
const ym = d.substring(0, 7);
if (!months[ym]) months[ym] = [];
months[ym].push(d);
});

const month_names = ["January","February","March","April","May","June","July","August","September","October","November","December"];

let calendars_html = "";
Object.keys(months).sort().forEach(ym => {
const [year, month] = ym.split("-").map(Number);
const first_day = new Date(year, month - 1, 1);
const last_day = new Date(year, month, 0);
const start_weekday = first_day.getDay();
const days_in_month = last_day.getDate();
const month_name = month_names[month - 1];

let cells = "";
for (let i = 0; i < start_weekday; i++) {
cells += "<td></td>";
}
for (let day = 1; day <= days_in_month; day++) {
const iso = `${year}-${String(month).padStart(2,'0')}-${String(day).padStart(2,'0')}`;
const is_unmarked = date_set.has(iso);
cells += `<td style="text-align:center; padding:6px; border-radius:4px;
${is_unmarked ? "background:#ffd1d1; color:#c0392b; font-weight:600;" : "color:#888;"}">${day}</td>`;
if ((start_weekday + day) % 7 === 0) cells += "</tr><tr>";
}

calendars_html += `
<div style="margin-bottom:18px;">
<div style="font-weight:600; margin-bottom:6px;">${month_name} ${year}</div>
<table style="width:100%; border-collapse:collapse; font-size:12px;">
<tr>
<td style="text-align:center; color:#999;">Su</td>
<td style="text-align:center; color:#999;">Mo</td>
<td style="text-align:center; color:#999;">Tu</td>
<td style="text-align:center; color:#999;">We</td>
<td style="text-align:center; color:#999;">Th</td>
<td style="text-align:center; color:#999;">Fr</td>
<td style="text-align:center; color:#999;">Sa</td>
</tr>
<tr>${cells}</tr>
</table>
</div>
`;
});

const d = new frappe.ui.Dialog({
title: `Unmarked Days — ${employee_name} (${dates.length} days)`,
size: "large",
fields: [
{
fieldtype: "HTML",
fieldname: "calendar_html",
options: `<div style="max-height:500px; overflow-y:auto; padding:5px;">${calendars_html}</div>`
}
]
});
d.show();
});
